File: LeetCode/random-pick-with-weight.cpp
// https://leetcode.com/problems/random-pick-with-weight/discuss/671460/Detailed-Explanation-or-Summary-of-2-different-solutions
// https://leetcode.com/problems/random-pick-with-weight/discuss/671550/Cumulative-Distribution-Function-or-Explanation-or-Binary-and-Linear-Search-code
class Solution
{
    vector<int> cdf;

public:
    Solution(vector<int> &w)
    {
        cdf = vector<int>(w.size());
        cdf[0] = w[0];
        for (int i = 1; i < cdf.size(); ++i)
        {
            cdf[i] = cdf[i - 1] + w[i];
        }
    }

    int pickIndex()
    {
        int x = cdf.back();
        int index = rand() % x;
        auto it = upper_bound(cdf.begin(), cdf.end(), index);
        return it - cdf.begin();
    }
};

File: LeetCode/random-pick-with-weight.java
// https://leetcode.com/problems/random-pick-with-weight/

// Binary Search on Prefix Sums array
class Solution {
  private int[] prefixSums;
  private int totalSum;

  public Solution(int[] w) {
    this.prefixSums = new int[w.length];

    int prefixSum = 0;
    for (int i = 0; i < w.length; ++i) {
      prefixSum += w[i];
      this.prefixSums[i] = prefixSum;
    }
    this.totalSum = prefixSum;
  }

  public int pickIndex() {
    double target = this.totalSum * Math.random();

    // run a binary search to find the target zone
    int low = 0, high = this.prefixSums.length;
    while (low < high) {
      // better to avoid the overflow
      int mid = low + (high - low) / 2;
      if (target > this.prefixSums[mid])
        low = mid + 1;
      else
        high = mid;
    }
    return low;
  }
}

// Linear Search Solution.
class Solution {
  private int[] prefixSums;
  private int totalSum;

  public Solution(int[] w) {
    this.prefixSums = new int[w.length];

    int prefixSum = 0;
    for (int i = 0; i < w.length; ++i) {
      prefixSum += w[i];
      this.prefixSums[i] = prefixSum;
    }
    this.totalSum = prefixSum;
  }

  public int pickIndex() {
    double target = this.totalSum * Math.random();
    int i = 0;
    // run a linear search to find the target zone
    for (; i < this.prefixSums.length; ++i) {
      if (target < this.prefixSums[i])
        return i;
    }
    // to have a return statement, though this should never happen.
    return -1;
  }
}
